DOC: Make explicit in pandas IO doc the imports and options

TanyaaCJain opened the pull request:
- [x] works on #28038 
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] Removes code header from IO doc page.
- [x] Uses explicit imports in the IO doc page of user guide on first occurrence of requirement.

Reference @datapythonista - [149](https://github.com/python-sprints/pandas-mentoring/issues/149)

TanyaaCJain committed:
DOC: Make explicit in pandas IO doc the iports and options

- Removes code header in IO doc page.
- Uses explicit imports in the IO page of user guide on first occurence of requirement.
- In reference to #28038

TanyaaCJain committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

TanyaaCJain committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into doc-remove-header

datapythonista commented:
Thanks for the work on this @TanyaaCJain, looks great. Did you check if the pandas and numpy options are being used? It'd probably be useful to generate a diff of the current (master) html generated for that page, and the one with this changes, and without the options. So we know for sure what's being affected by this change.

@TomAugspurger @jorisvandenbossche @jreback this is the option that makes more sense to me. I think having a block with the imports and the options at the top of every file doesn't really look great. Thought?

TomAugspurger commented:
This (suppressing the imports) seems fine for rendered HTML on the majority of the pages. I think we intentionally don't suppress them on the first few of the getting started.

For the binder version, presumably these will be visible / executed? Or are we not thinking that far ahead yet?

datapythonista commented:
The main motivation of this change is not having hidden code, so the integration with Binder is easier and there is no magic for the users.

For the API pages I guess it will be more controversial. But in my opinion the changes to the user guide are very small and surely worth.

jorisvandenbossche commented:
> The main motivation of this change is not having hidden code,

Note that in the current diff of this PR, it is still "hidden" to the user for the online html docs

datapythonista commented on doc/source/user_guide/io.rst:
```diff
@@ -2,12 +2,11 @@
+   import pandas as pd
+   pd.options.display.max_rows = 15
    clipdf = pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6], 'C': ['p', 'q', 'r']},
                          index=['x', 'y', 'z'])
```
Didn't realize this was hidden, thanks @jorisvandenbossche for pointing it out.

@TanyaaCJain this block is not shown to the user, can you move the `clipdf` data where it's first used, and the import to the first next block (assuming there are no more hidden blocks before.

TanyaaCJain replied:
Made the changes in the commit 18a9b7c.

TanyaaCJain committed:
DOC: Update io.rst to explicitly add imports and options

- imported pandas and added its options in the first visible code block
- moved the suppressed or invisible code block using clipdf just before the code block using the clipdf variable
- imported `BytesIO` in its first occurrence of requirement unlike previously imported along with `StringIO`
- imported os again for adding it on its first occurrence of requirement in a visible code block and is earlier imported in a suppressed code block.

TanyaaCJain committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into doc-remove-header

TanyaaCJain commented:
The new commit does this:
- [x] imported `pandas` and added its options in the first visible code block
- [x] moved the suppressed or invisible code block using `clipdf` just before the code block using the `clipdf` variable
- [x] imported `BytesIO` in its first occurrence of requirement unlike previously imported along with `StringIO`
- [x] imported `os` again for adding it on its first occurrence of requirement in a visible code block and is earlier imported in a suppressed code block.

I need help with:
- `np.random.seed(123456)` is currently (in this PR) being called just before the first occurrence of `np.random.randn()` which is in a suppressed block. What is to be done?
   - Add it again in a visible code block with first occurrence of `np.random.randn()` as done with `import os`?
   - Let it only be in the suppressed code block?
- I am not able to get any diff while comparing the current (master) html with the one in this PR. Can you help me with the code. Probably, not writing the right one.

Please suggest if there are any other changes as well.

jorisvandenbossche commented on doc/source/user_guide/io.rst:
```diff
@@ -137,7 +128,9 @@ usecols : list-like or callable, default ``None``
 
-     from io import StringIO, BytesIO
+     import pandas as pd
+     pd.options.display.max_rows = 15
```
I think you can actually remove this line. 

I quickly checked and there are only a few longer dataframes shown in this page, and they are all longer. So with the current default of only showing first/last 5 rows when truncated, this option has no effect (it would only have the effect that dataframes longer than 15 but shorter than 60 would still be truncated, but such dataframes are not present in this file)

TanyaaCJain replied:
Made the changes in commit d3036f3.

jorisvandenbossche commented:
> np.random.seed(123456)

In general, do we care much about setting this?

TanyaaCJain committed:
DOC: Removed pandas row display option because it is not required at the moment

TanyaaCJain committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into doc-remove-header

datapythonista commented on doc/source/user_guide/io.rst:
```diff
@@ -3266,6 +3261,12 @@ clipboard (CTRL-C on many operating systems):
+   :suppress:
+
+   clipdf = pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6], 'C': ['p', 'q', 'r']},
+                         index=['x', 'y', 'z'])
```
I'm not sure if this is really needed. Do you mind having a look @TanyaaCJain

TanyaaCJain replied:
Oh yes.. I missed mentioning this. Thanks @datapythonista for the reminder! Even I couldn't find this assignment be of use anywhere in the file. So, should we go ahead with getting rid of this?

datapythonista replied:
I have he feeling that the `read_clipboard` was an `ipython` block before, and we had this `clipdf` and a `to_clipboard` before reading from the clipboard. But since not it's a python block and the code is written manually, I guess we can simply get rid of this.

datapythonista reviewed (commented):
I don't have a preference regarding the seed. I think for the long term would be nice to avoid examples with random data. But for now I think it's perfectly fine to remove the seed and have different examples every time (I don't think anybody cares, including search engines, and we the production docs are frozen between releases).

It'd also be ok to me if we decide to remove the numpy printooptions. I think data looks nicer with it, but if setting it can cause confusion to users, not a problem for me to get rid of it.

TanyaaCJain commented:
> * I am not able to get any diff while comparing the current (master) html with the one in this PR. Can you help me with the code. Probably, not writing the right one.

Also, @datapythonista can you help me with this?

datapythonista commented:
> Also, @datapythonista can you help me with this?

My idea was to go to an updated master branch `git checkout master && git fetch upstream && git merge upstream/master`, then generate the documentation (just this page if you want to save time) `./doc/make.py --single=user_guide/io.rst` and then save somewhere the resulting html file `doc/build/html/user_guide/io.html`.

The, go to your branch, generate that page again and do a `diff -u io_from_master.html io_without_header.html`.

That should highlight everything that will be different for the user. Better to do it wihout removing the seed, otherwise all the data will be different.

Does that make sense?

TanyaaCJain commented:
This is the diff generated, can be viewed as html if the extension is changed.
[diff.txt](https://github.com/pandas-dev/pandas/files/3535388/diff.txt)
I'll commit the new changes after this.

TanyaaCJain committed:
DOC: Remove unused clipdf assignment, numpy print options and seed

TanyaaCJain committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into doc-remove-header

TanyaaCJain committed:
DOC: Remove unused clipdf assignment, numpy print options and seed

datapythonista commented:
didn't think about that, but the changes make the number of the `In [x]` and `Out [x]` of jupyter change, and the diff is huge, not so easy to see what changed. But thank you for it.

datapythonista commented on doc/source/user_guide/io.rst:
```diff
@@ -137,7 +128,8 @@ usecols : list-like or callable, default ``None``
   .. ipython:: python
 
-     from io import StringIO, BytesIO
+     import pandas as pd
```
I don't think that's an assumption. It is implicitly being proposed, for simplicity and for inertia on how the header worked, but it's surely open to discussion. I thought about that too, and I was also wondering if it'd make sense to split the long pages we have now in the user guide in shorter pages. I guess that could make things easier if we finally make the code runnable.

datapythonista commented:
Answering this comment from @WillAyd on the issue: https://github.com/pandas-dev/pandas/issues/28038#issuecomment-528432981

This is still open to discussion, but personally I much like this approach than the header. There are some advantages:
- Simpler
- Runnable code with no magic for users
- If we make the code runnable with binder, not having headers/hidden code will probably help
- Sphinx will report the correct line on errors
- If what we did in this PR can extrapolate to the rest, there is not so much needed in the header anyway

I agree with your point that in a huge documentation page, having the import just at the beginning doesn't necessarily make sense, the document will often not be accessed sequentially. But discussed with @jorisvandenbossche, and seems reasonable to split very long pages. And with binder in mind, probably worth having the imports repeated at the beginning of main sections, rather than hidden.

So, while this PR may not be a perfect solution yet, I think it's a step forward in the right direction.

TomAugspurger commented:
Agreed with Marc.

On Tue, Sep 10, 2019 at 10:41 AM Marc Garcia <notifications@github.com>
wrote:

> Answering this comment from @WillAyd <https://github.com/WillAyd> on the
> issue: #28038 (comment)
> <https://github.com/pandas-dev/pandas/issues/28038#issuecomment-528432981>
>
> This is still open to discussion, but personally I much like this approach
> than the header. There are some advantages:
>
>    - Simpler
>    - Runnable code with no magic for users
>    - If we make the code runnable with binder, not having headers/hidden
>    code will probably help
>    - Sphinx will report the correct line on errors
>    - If what we did in this PR can extrapolate to the rest, there is not
>    so much needed in the header anyway
>
> I agree with your point that in a huge documentation page, having the
> import just at the beginning doesn't necessarily make sense, the document
> will often not be accessed sequentially. But discussed with
> @jorisvandenbossche <https://github.com/jorisvandenbossche>, and seems
> reasonable to split very long pages. And with binder in mind, probably
> worth having the imports repeated at the beginning of main sections, rather
> than hidden.
>
> So, while this PR may not be a perfect solution yet, I think it's a step
> forward in the right direction.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/28089?email_source=notifications&email_token=AAKAOIXHWWUQ3ALT6M46GLLQI65YZA5CNFSM4IOU7FTKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6LRWUQ#issuecomment-529996626>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAOIQIAFX4ZXEVNEXKWOTQI65YZANCNFSM4IOU7FTA>
> .
>

TanyaaCJain commented:
Would you all prefer to split up this file into smaller sections too or just want it for way more lengthier ones? Though I think splitting doc into sections would help to understand difference in file much easily, if preferred.


> didn't think about that, but the changes make the number of the `In [x]` and `Out [x]` of jupyter change, and the diff is huge, not so easy to see what changed. But thank you for it.

I tried for a few other ways that could possibly generate a difference ignoring the ipython lines but none generated the desired output. I’ll update if I do find one.

WillAyd commented:
If runnable code with binder is the main draw here (that's what I am inferencing at least, perhaps incorrectly) is there any reason why we don't defer moving things around like this until something like binder would actually get implemented?

TomAugspurger commented:
Not sure where Marc is at on that, but I stopped my prototype at
https://github.com/TomAugspurger/pandas-binder when I noticed that the `{{
header }}` stuff wasn't working. So from that attempt's POV, this is a
blocker.

On Tue, Sep 10, 2019 at 2:58 PM William Ayd <notifications@github.com>
wrote:

> If runnable code with binder is the main draw here (that's what I am
> inferencing at least, perhaps incorrectly) is there any reason why we don't
> defer moving things around like this until something like binder would
> actually get implemented?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/28089?email_source=notifications&email_token=AAKAOIT3J3HUHTBE6H4BZYDQI735VA5CNFSM4IOU7FTKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6MJYLQ#issuecomment-530095150>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAOIRBFLK5GSV6QPQWRCLQI735VANCNFSM4IOU7FTA>
> .
>

WillAyd commented:
Sounds good. I'll defer to you all then as I think you have a better big picture view of this and I don't think anything here is a showstopper. Just wanted to offer an alternative view - thanks!

jorisvandenbossche commented:
@WillAyd to come back to your concern (as it was not fully addressed IMO):

> I'm not sure I am on board with the concept. Associating each import with its first reference in the rst feels a little arbitrary and I think is just confusing. Seems like it could be somewhat of a nuisance when moving things around; nothing major but in any case a step backwards from just knowing standard imports are done in the header

Currently, all standard imports (pandas, numpy) are at the beginning of each rst file, but hidden for the user. Now we still do them at the top of the file, but visible. So a net improvement, and not arbitrary.

Or at least, that's the case for the pandas import. But you are right that if the `from io import BytesIO` import is somewhere in the middle of the file, and then re-used without import near the end of the file in a totally different section, this can be confusing (not necessarily more confusing than not showing the import at all as it is now, though):

- for writing docs, I don't think this is a problem, we have CI checking for this now to prevent errors
- for really common imports (I think mainly `numpy`), it's maybe worth to still put this at the top of the file with the pandas import instead at the location of first usage? (in the end, that is also what I do in my notebook workflow)
- for less common imports that we want to do at the place of usage, I think we can actually repeat this import multiple times if they happen in a single file but in unrelated sections. 

And as Marc said, going to smaller files will also help with this.

jorisvandenbossche commented:
> Would you all prefer to split up this file into smaller sections too

@TanyaaCJain I think we want to work towards that, but that's certainly for another PR (for the IO page we have https://github.com/pandas-dev/pandas/issues/10446 about this)

jreback commented:
status on this PR @jorisvandenbossche @datapythonista

datapythonista commented:
If nobody objects to remove the header in the docs, the seeds, and make the imports explicit, this is ready to be merged, and we can start working on the rest of the docs.

jorisvandenbossche commented:
I am fully behind removing the header and moving the imports in actual code-blocks.

One thing that might be good to decide on before starting other PRs on the rest of the docs, is where to put those imports.  
In the current PR, it is "the first code block where it is used" for everything. 

I think that is fine for less common imports, but it could be an option to still put eg numpy and pandas (and matplotlib.pyplot for those rst files that use it) in a visible code block in the beginning of each file. 

(but also fine to first try out with the approach as in this PR, and evaluate later about that question)

WillAyd commented:
@datapythonista @jorisvandenbossche still want to merge?

TomAugspurger commented:
Yeah, I think so.

Another point, maybe already mentioned, is that explicit imports will fix
the line-number reported by sphinx in errors.

On Thu, Nov 7, 2019 at 3:11 PM William Ayd <notifications@github.com> wrote:

> @datapythonista <https://github.com/datapythonista> @jorisvandenbossche
> <https://github.com/jorisvandenbossche> still want to merge?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/28089?email_source=notifications&email_token=AAKAOIV7TW5CQC5DA6CRXS3QSSAABA5CNFSM4IOU7FTKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDN2M4Y#issuecomment-551265907>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKAOIXHK67JQH2H4M4JQT3QSSAABANCNFSM4IOU7FTA>
> .
>

WillAyd commented:
Thanks @TanyaaCJain
